### PR TITLE
matchers for less than and greater than

### PR DESCRIPTION
**Pull request by dylan-conway**

### CODE REVIEW

File: src/bun.js/bindings/bindings.cpp [hunk 1]
```diff
@@ -2350,6 +2350,52 @@ int64_t JSC__JSValue__toInt64(JSC__JSValue val)
     return _val.asAnyInt();
 }
 
+uint8_t JSC__JSValue__asBigIntCompare(JSC__JSValue JSValue0, JSC__JSGlobalObject* globalObject, JSC__JSValue JSValue1)
+{
+    JSValue v1 = JSValue::decode(JSValue0);
+    JSValue v2 = JSValue::decode(JSValue1);
+    ASSERT(v1.isHeapBigInt() || v1.isBigInt32());
+
+#if USE(BIGINT32)
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/bindings.cpp:2359]: context: BIGINT32 is for 32 bit platforms I think

File: src/bun.js/test/jest.zig [hunk 2]
```diff
@@ -828,6 +828,258 @@ pub const Expect = struct {
...
+        const lesser_value = arguments[0];
+        lesser_value.ensureStillAlive();
+
+        const value = Expect.capturedValueGetCached(thisValue) orelse {
+            globalObject.throw("Internal consistency error: thie expect(value) was garbage collected but it should not have been!", .{});
+            return .zero;
+        };
+        value.ensureStillAlive();
+
+        if ((!value.isNumber() and !value.isBigInt()) or (!lesser_value.isNumber() and !lesser_value.isBigInt())) {
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig]: We need to call `JSC::toNumber` here

See this https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/toPrimitive

File: src/bun.js/test/jest.zig [hunk 3]
```diff
@@ -828,6 +828,258 @@ pub const Expect = struct {
...
+            return .zero;
+        }
+
+        const not = this.op.contains(.not);
+        var pass = false;
+
+        if (!value.isBigInt() and !lesser_value.isBigInt()) {
+            pass = value.asNumber() >= lesser_value.asNumber();
+        } else if (value.isBigInt()) {
+            switch (value.asBigIntCompare(globalObject, lesser_value)) {
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig]: switch can be expressions too

```zig
    pass = switch (value.asBigIntCompare(globalObject, lesser_value)) {
         .greater_than, .equal => true,
         else => pass,
    };
```

File: src/bun.js/test/jest.zig [hunk 4]
```diff
@@ -828,6 +828,258 @@ pub const Expect = struct {
...
+        const lesser_value = arguments[0];
+        lesser_value.ensureStillAlive();
+
+        const value = Expect.capturedValueGetCached(thisValue) orelse {
+            globalObject.throw("Internal consistency error: thie expect(value) was garbage collected but it should not have been!", .{});
+            return .zero;
+        };
+        value.ensureStillAlive();
+
+        if ((!value.isNumber() and !value.isBigInt()) or (!lesser_value.isNumber() and !lesser_value.isBigInt())) {
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig]: ditto

File: src/bun.js/test/jest.zig [hunk 5]
```diff
@@ -828,6 +828,258 @@ pub const Expect = struct {
...
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig]: ditto

File: src/bun.js/test/jest.zig [hunk 6]
```diff
@@ -828,6 +828,258 @@ pub const Expect = struct {
...
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig]: ditto

File: src/bun.js/test/jest.zig (hunk 3 above)
Comment (dylan-conway) [reply src/bun.js/test/jest.zig]: added switch expressions

File: src/bun.js/test/jest.zig [hunk 7]
```diff
@@ -828,6 +828,258 @@ pub const Expect = struct {
...
+        if ((!value.isNumber() and !value.isBigInt()) or (!other_value.isNumber() and !other_value.isBigInt())) {
+            globalObject.throw("Expected and actual values must be numbers or bigints", .{});
+            return .zero;
+        }
+
+        const not = this.op.contains(.not);
+        var pass = false;
+
+        if (!value.isBigInt() and !other_value.isBigInt()) {
+            pass = value.asNumber() > other_value.asNumber();
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig:868]: This needs to be `toNumber()`

`asNumber` does no coercion

`toNumber` coerces

Comment (Jarred-Sumner) [reply src/bun.js/test/jest.zig:868]: or does jest not coerce numbers?

Comment (dylan-conway) [reply src/bun.js/test/jest.zig:868]: Jest doesn't coerce. Their matchers for these only accept numbers or bitints

**Pull request merged**